False-positive: Starting from version 4.0.0 false-positive on pytest-bdd fixtures

### Bug description

We have a set of `pytest_bdd` based tests, where every step is decorated with its description and has the same name `_`. Up until version 4.0.0, it was not a problem. But now every single step description after the first one emits a failed test `function-redefined`. Nothing was changed in a code and if I

### Configuration

```ini

```

### Command used

```shell
± uv run pylint -j 0 example.py
```

### Pylint output

```python
************* Module example
example.py:15:0: E0102: function already defined line 10 (function-redefined)
example.py:20:0: E0102: function already defined line 10 (function-redefined)

------------------------------------------------------------------
Your code has been rated at 0.00/10 (previous run: 0.00/10, +0.00)
```

### Expected behavior

Same as it was in versions before 4.0.0:
```
************* Module example
example.py:15:0: E0102: function already defined line 10 (function-redefined)
example.py:20:0: E0102: function already defined line 10 (function-redefined)

------------------------------------------------------------------
Your code has been rated at 0.00/10 (previous run: 0.00/10, +0.00)
```

This is an example for pytest version 3.3.9 on the very same file. File itself:

[example.py](https://github.com/user-attachments/files/22904683/example.py)

.

### Pylint version

```shell
debian:
± uv run  pylint --version
pylint 3.3.9
astroid 3.3.11
Python 3.13.5 (main, Jun 25 2025, 18:55:22) [GCC 14.2.0]

Same on Ubuntu:
± pylint --version
pylint 4.0.0
astroid 4.0.1
Python 3.13.5 (main, Jun 12 2025, 08:56:14) [GCC 13.3.0]
```

### OS / Environment

Debian:
No LSB modules are available.
Distributor ID: Debian
Description:    Debian GNU/Linux 13 (trixie)
Release:        13
Codename:       trixie

Ubuntu:
± lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 24.04.2 LTS
Release:        24.04
Codename:       noble

### Additional dependencies

```python

```

--- a/doc/whatsnew/fragments/10665.false_positive
+++ b/doc/whatsnew/fragments/10665.false_positive
@@ -1,0 +1,7 @@
+Fix a false positive for ``function-redefined`` (E0102) when reusing names that
+match ``dummy-variables-rgx`` (such as ``_``), which is common for
+``pytest-bdd`` step definitions. This restores the behavior from before pylint
+4.0.0; as a consequence the false negative fixed in #9894 is reintroduced for
+functions whose name matches ``dummy-variables-rgx``.
+
+Closes #10665

--- a/pylint/checkers/base/basic_error_checker.py
+++ b/pylint/checkers/base/basic_error_checker.py
@@ -640,6 +640,9 @@ class BasicErrorChecker(_BasicChecker):
                     ):
                         return
 
+            dummy_variables_rgx = self.linter.config.dummy_variables_rgx
+            if dummy_variables_rgx and dummy_variables_rgx.match(node.name):
+                return
             self.add_message(
                 "function-redefined",
                 node=node,

--- a/tests/functional/f/function_redefined.py
+++ b/tests/functional/f/function_redefined.py
@@ -94,7 +94,7 @@ def math(): # [function-redefined]
     pass
 
 import math as _
-def fun():
+def _():
     pass
 
 # pylint: disable=too-few-public-methods

--- a/tests/functional/f/function_redefined_private_function.py
+++ b/tests/functional/f/function_redefined_private_function.py
@@ -1,6 +1,14 @@
 # pylint: disable=missing-module-docstring, missing-function-docstring
+# Names matching ``dummy-variables-rgx`` are exempt from function-redefined,
+# e.g. ``_`` used repeatedly as a pytest-bdd step name (issue #10665).
+def _():
+    pass
+
+def _():
+    pass
+
 def _my_func():
     pass
 
-def _my_func():  # [function-redefined]
+def _my_func():
     pass

--- a/tests/functional/f/function_redefined_private_function.txt
+++ b/tests/functional/f/function_redefined_private_function.txt
@@ -1,1 +1,0 @@
-function-redefined:5:0:5:12:_my_func:function already defined line 2:UNDEFINED

